stumpy/scrump.py

```diff
@@ -103,67 +109,125 @@ def _compute_PI(
     for i in indices[start:stop]:
         Q = T_A[i : i + m]
         QT[:] = core._sliding_dot_product(Q, T_B)
-        # Update P[i] relative to all T[j : j + m]
         squared_distance_profile[:] = core._mass(Q, T_B, QT, μ_Q[i], σ_Q[i], M_T, Σ_T)
         squared_distance_profile[:] = np.square(squared_distance_profile)
         if excl_zone is not None:
-            zone_start = max(0, i - excl_zone)
-            zone_stop = min(w, i + excl_zone)
-            squared_distance_profile[zone_start : zone_stop + 1] = np.inf
-
-            # only for self-join
-            mask = squared_distance_profile < P_squared[thread_idx]
-            P_squared[thread_idx][mask] = squared_distance_profile[mask]
-            I[thread_idx][mask] = i
-
-        I[thread_idx, i] = np.argmin(squared_distance_profile)
-        P_squared[thread_idx, i] = squared_distance_profile[I[thread_idx, i]]
-        if P_squared[thread_idx, i] == np.inf:  # pragma: no cover
-            I[thread_idx, i] = -1
-        else:
-            j = I[thread_idx, i]
-            # Given the squared distance, work backwards and compute QT
-            QT_j = (m - P_squared[thread_idx, i] / 2.0) * (Σ_T[j] * σ_Q[i]) + (
-                m * M_T[j] * μ_Q[i]
+            core._apply_exclusion_zone(squared_distance_profile, i, excl_zone, np.inf)
+
+        nn_idx = np.argmin(squared_distance_profile)
+        if nn_idx not in I[thread_idx, i]:
+            # It is more than likely that the top-k values for the `i`-th subsequence
+            # will be already populated. So, we must shift-insert here
```

Comment:
Is this comment informative enough?

Reply:
I think we need a just a little bit more. While the context is still fresh in my head, I understand this comment but I think something like:

```
# Since the top-k values for the `i`-th subsequence may already
# be updated/populated (i.e., not all values in `I[thread_idx]` are
# equal to `-1` or not all values in `P_squared[thread_idx, i]` are
# equal  to `np.inf`) in other previous iterations, we must shift-insert
# here rather than assign values to the first element
```

It's quite verbose but it directly addresses where my brain went to the first time I saw this. That is, I had thought that `P_squared` and `I` must always be filled with `np.inf` and `-1` and is unaffected by other iterations of `i`. As you pointed out, this is `False`. And this comment stops/short-circuits/discredits that thinking immediately. What do you think?

Reply:
Yeah... I think it is better to go with the longer message you suggested. As you mentioned, we know these stuff right now, but we may forget the purpose of doing shift/insert later. (In fact, I myself made the same mistake while I was writing the code. So, there is a high chance that we may get into that trap later if we do not provide a good comment.

(Just for the records, note that even if  `P_squared[thread_idx, i]` / `I[thread_idx, i]` was  previously updated only once, that one element will be at `index 0`.)